Detect and handle NCCL errors appropriately in ProcessGroupNCCL.

pritamdamania87 opened the pull request:
Summary:
This change adds the following functionality:
1) WorkNCCL isCompleted, isSuccess methods check for NCCL errors and set the
appropriate exception.
2) Added a watchdog thread to ProcessGroupNCCL which checks for errors in the
cached communicators and removes them from the cache.
3) Use ncclCommAbort in NCCLComm destructor since ncclCommDestroy can block
forever waiting for work.
4) Added a simulate_nccl_errors.py script to simulate NCCL errors.

https://github.com/pytorch/pytorch/issues/17882

Differential Revision: D16220638

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -176,11 +209,72 @@ ProcessGroupNCCL::ProcessGroupNCCL(
+
+    std::unique_lock<std::mutex> lock(watchdogCVMutex_);
+    watchdogCV_.wait_for(lock, kWatchdogThreadSleep, [&]() -> bool {
+      return terminateWatchdog_;
```
Shall we use `.load()` here just to stay consistent with above?

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -271,6 +369,9 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
   C10D_NCCL_CHECK(ncclGroupEnd());
 
+  // Hold the lock before modifying the cache.
+  std::lock_guard<std::mutex> lock(devNCCLCommMapLock_);
```
If this is only for `devNCCLCommMap_`, is it OK to move `ncclStreams_` and `ncclEvents_` out of the scope?

pietern commented:
@pritamdamania87 There are some NCCL build issues here. Looks like the function prototypes for `nccl
CommAbort` and friends are not available. Perhaps they are gated by some macro? Or we're on a version older than 2.4 for these CI base images...

pritamdamania87 commented:
@pietern Where do you see the NCCL build issues? Looking at the builds that failed, it seems to be because I was using std::chrono_literals which isn't supported in C++11.

pietern commented:
@pritamdamania87 It's a different error now. Looks like gtest isn't linked to `ProcessGroupNCCLTestErrors` per the undefined symbols errors.

pritamdamania87 commented:
@pietern Fixed the gtest issue and all tests are passing now.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -163,6 +181,16 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
+    if (exception()) {
+      std::rethrow_exception(exception());
+    }
+  }
```
Above, `isSuccess` is defined as "there is no exception on the communicator AND the events associated with kernel completion are ready". This means the path here is hit if "there is an exception on the communicator OR the events are not yet ready" (i.e. the kernel has yet to run or is still running). Then it checks for exceptions on the communicator and throws it if there is one. But, if the second part of the conditional is true, then there will be no exception, and the function returns. If after that the communicator still fails, we have a problem. The `wait()` function that the user called has returned and new CUDA kernels will have been queued to run, but there is no way to raise the yet-to-happen exception on a user thread.

Moreover, the watchdog will abort the communicator, and work scheduled after it will run, assuming the operation completed successfully. A worst case scenario here is that an optimizer runs on garbage gradients, the user saves a checkpoint, figures out an error happened on the next iteration, restarts, and restores from a garbage checkpoint.

At the moment, I don't think there is a way we can avoid doing full on synchronization on completion here.

pritamdamania87 replied:
If the user really wants to check for all errors, they need to do something like this:

```
work = pg.allreduce(tensors)
pg.barrier()->wait()
if (!work.isSuccess()) {
  // handle error.
}
```

My feeling is that the decision on whether or not to continue other operations without checking if previous operations have succeeded or failed is on the user. If the user decides to use the result of an operation which has failed (probably garbage gradients) without checking its status, its on them.

Another option the user has is to batch a lot of operations together to run asynchronously and then have a single barrier to wait for completion and check for errors.

> Moreover, the watchdog will abort the communicator, and work scheduled after it will run, assuming the operation completed successfully. A worst case scenario here is that an optimizer runs on garbage gradients, the user saves a checkpoint, figures out an error happened on the next iteration, restarts, and restores from a garbage checkpoint.

In this case, doesn't the user need to have a barrier before running the optimizer (to ensure the previous operation completed?). If so, the user should check for the status of the previous operation after the barrier.

pietern replied:
Just to clarify, my comment here was w.r.t. the use of `isSuccess` inside `wait`, not the definition of `isSuccess` by itself. The current use of `isSuccess` allows for `wait` to return without actually checking for completion of the operation.

> My feeling is that the decision on whether or not to continue other operations without checking if previous operations have succeeded or failed is on the user. If the user decides to use the result of an operation which has failed (probably garbage gradients) without checking its status, its on them.

This is counter-intuitive w.r.t. the current behavior. Today, if you call `wait()`, either the subsequent operations will execute on the real result of the operation, or your processes will hang indefinitely. Adding an in-between mode where the subsequent operations execute, but they execute on garbage, will lead to unexpected results for users. Therefore I strongly feel that `wait` must synchronize. 

> Another option the user has is to batch a lot of operations together to run asynchronously and then have a single barrier to wait for completion and check for errors.

This is what DDP does as well -- first it kicks off all the work it needs to do, asynchronously, and then when it's done with autograd, it waits on all of them in order. In practice, most of them will have already completed in the background, so the synchronization won't involve actually waiting for something to complete.

> In this case, doesn't the user need to have a barrier before running the optimizer (to ensure the previous operation completed?). If so, the user should check for the status of the previous operation after the barrier.

This shouldn't be a separate operation, or we open ourselves up to the unexpected behavior. It should be all part of `wait` in my opinion. There is also no need for a barrier, if `wait` is indeed synchronous.

pritamdamania87 replied:
>This is counter-intuitive w.r.t. the current behavior. Today, if you call wait(), either the subsequent operations will execute on the real result of the operation, or your processes will hang indefinitely. Adding an in-between mode where the subsequent operations execute, but they execute on garbage, will lead to unexpected results for users. Therefore I strongly feel that wait must synchronize.

Will update the PR to make `wait()` synchronous. Although, doesn't this change the behavior for current users? The operation was non-blocking and now we are making it blocking. Is there any way to inform users about this change apart from adding documentation to the method?

mrshenli replied:
Would I be correct if I assume the above questions are already resolved after introducing `NCCL_BLOCKING_WAIT ` env var?

pietern reviewed (commented):
Besides the correctness here, I think we need some way to make this optional. We can turn it on by default, but we'd retain the ability to turn it off by setting an environment variable, for example. This is just in case that this causes issues for folks, either in terms of crashes, or in terms of performance.

pietern commented on test/simulate_nccl_errors.py:
```diff
@@ -0,0 +1,38 @@
+        logging.info('Second allreduce successful: {}'.format(work.is_success()))
+    else:
+        logging.info('Aborting all other ranks.')
+        os.abort()
```
What if you launch 2 processes, but run the collective on only one of them. Then there are two paths to test, 1) the other process is killed and we expect to see an error exception, and 2) the other process hangs around and we expect to see a timeout exception.

pritamdamania87 replied:
I tried a lot of different scenarios including the two that you mentioned, but we don't see any errors on the communicator and the unit test just times out. That is the reason why I actually came up with this script and ran it across two different hosts to try and simulate an error. I also tried running this script on the same host with two different processes and I didn't see any exceptions and rank 0 just blocked forever.

pietern replied:
Hmm, that is quite concerning. If we can't detect a local crash there are many situations where the error reporting won't kick in and the remaining processes will hang.

Was this where you had only a single process call the collective and the other one sleep (i.e. not crash), or the other one crash? Either situation should be handled (i.e. if you send a SIGSTOP to the other process), but if it doesn't detect a local crash that would be a surprise.

pritamdamania87 replied:
> Was this where you had only a single process call the collective and the other one sleep (i.e. not crash), or the other one crash?

I didn't have the other one sleep, I either cleanly exited the other process or aborted it with SIGABRT. Although, in the latest version of the PR I have added a timeout which should throw an exception if something like this happens. I've also added several test cases to `test_c10d.py` which test various different patterns of sleep/crash for the other process.

pritamdamania87 commented:
> Besides the correctness here, I think we need some way to make this optional. We can turn it on by default, but we'd retain the ability to turn it off by setting an environment variable, for example. This is just in case that this causes issues for folks, either in terms of crashes, or in terms of performance.

I understand making a feature optional if its something we are experimenting with and maybe is a perf optimization that might have side effects. But I feel error checking is extremely critical and we shouldn't have a way to turn it off. This might cause other problems in the future where users would just disable error checking for convenience/if they don't want to deal with errors.

pritamdamania87 commented:
@pietern Looks like the lint issue is for a file that I didn't touch. Is this expected? I thought we ran lint only for files changed. Also, a couple of other builds failed because they were aborted. Is there anyway to re-run just those builds instead of running the whole suite again?

mrshenli commented:
> Looks like the lint issue is for a file that I didn't touch. Is this expected? I thought we ran lint only for files changed. 

If this PR does not touch those files, it's OK to ignore the lint failure. 

> Also, a couple of other builds failed because they were aborted. Is there anyway to re-run just those builds instead of running the whole suite again?

@pytorchbot retest this please

mrshenli commented on test/test_c10d.py:
```diff
@@ -2831,6 +2848,83 @@ def _test_broadcast_coalesced(self, process_group, device):
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+        process_group.allreduce(torch.rand(10).cuda(self.rank))
+        if self.rank == 0:
+            # Should run fine without raising any errors.
```
Let's be more explicit in the comment. IIUC, shall we add something like: This `allreduce` does not block Python thread as `allreduce` enqueues the cuda operation, and then `wait` only blocks the current cuda stream.

mrshenli commented on test/test_c10d.py:
```diff
@@ -2831,6 +2848,83 @@ def _test_broadcast_coalesced(self, process_group, device):
+            work = process_group.allreduce(torch.rand(10).cuda(self.rank))
+            work.wait()
+
+            # Now the work scheduled next should hang forever since the previous
```
Curious, which line does it hang at? Is it exactly [here](https://github.com/jeffdaily/pytorch/blob/f43b725472406ec1cfb35fe82dd3cc93ed05fcf7/torch/lib/c10d/ProcessGroupNCCL.cpp#L434) when calling `ncclAllReduce`? If so, it seems it does not only enqueue the op, but also perform some checking on the `NcclComm`.

mrshenli commented on test/test_c10d.py:
```diff
@@ -2831,6 +2848,83 @@ def _test_broadcast_coalesced(self, process_group, device):
+            # allreduce will never complete.
+            t = threading.Thread(target=self._run_all_reduce, args=(process_group,))
+            t.start()
+            t.join(int(get_timeout(self.id()) / 2))
```
Why divided by 2 there? Is it just to make sure that join returns before the entire test timeout?

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -163,6 +194,23 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
   }
+
+  // In case of blocking, wait for the operation to complete.
+  if (blockingWait_) {
```
similar code might need to be added to `AsyncAllreduceCUDAWork` in `ProcessGroupGloo`. It can come in a separate PR.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -194,11 +266,81 @@ ProcessGroupNCCL::ProcessGroupNCCL(
   processGroupID_ = std::to_string(processGroupCounterMap_[groupKey]);
   groupPgID_ = groupName_ + "_" + processGroupID_;
   pgUniqueNCCLIDCnt_[groupPgID_] = -1;
+  ncclCommWatchdogThread_ =
```
I am debating with myself on whether this should be optional and whether we should allow users to control this. If an application is never bothered by NCCL comm errors, will this still be useful? If not, does it make sense to provide an arg to avoid always having an additional thread?

mrshenli commented on torch/lib/c10d/test/CMakeLists.txt:
```diff
@@ -23,6 +23,7 @@ if(USE_CUDA)
   endif()
   if(USE_C10D_NCCL)
     c10d_add_test(ProcessGroupNCCLTest.cpp c10d c10d_cuda_test)
+    c10d_add_test(ProcessGroupNCCLErrorsTest.cpp c10d c10d_cuda_test gtest_main)
```
Why this is added to a larger scope than `ProcessGroupNCCLTest.cpp`?

mrshenli reviewed (commented):
The inline comments are mostly nits. The main concern is that we cannot detect errors in single-server use cases. 

@mcarilli @ngimel could you please help take a look at this PR as well? Thanks!

mrshenli reviewed (approved):
LGTM! Before landing, shall we run some micro-benchmark to make sure there is no perf regression?

The concern still remains as we cannot detect failures on a single server. However, in terms of failure detection, this is not a regressions, so I think we can land this one, and address single server case in followup PRs.

mrshenli commented:
#24906 for tracking purpose.

pritamdamania committed:
Detect and handle NCCL errors appropriately in ProcessGroupNCCL. (#22907)

Summary:
This change adds the following functionality:
1) WorkNCCL isCompleted, isSuccess methods check for NCCL errors and set the
appropriate exception.
2) Added a watchdog thread to ProcessGroupNCCL which checks for errors in the
cached communicators and removes them from the cache.
3) Use ncclCommAbort in NCCLComm destructor since ncclCommDestroy can block
forever waiting for work.
4) Added a simulate_nccl_errors.py script to simulate NCCL errors.

https://github.com/pytorch/pytorch/issues/17882
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22907

Test Plan: 1) Run the simulate_nccl_errors.py to verify NCCL errors are caught.

Differential Revision: D16220638

fbshipit-source-id: 0a44151adf5bfcde02fec1efd7dd895afbfd2584

suo commented:
Looks like this PR broke master: https://circleci.com/gh/pytorch/pytorch/2527700?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

```
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:114:37: error: redeclaration 'c10d::ProcessGroupNCCL::kWatchdogThreadSleepMillis' differs in 'constexpr'
Aug 21 04:05:34  constexpr int64_t ProcessGroupNCCL::kWatchdogThreadSleepMillis = 100;
Aug 21 04:05:34                                      ^
Aug 21 04:05:34 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:1:0:
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/ProcessGroupNCCL.hpp:253:24: error: from previous declaration 'c10d::ProcessGroupNCCL::kWatchdogThreadSleepMillis'
Aug 21 04:05:34    static const int64_t kWatchdogThreadSleepMillis;
Aug 21 04:05:34                         ^
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:114:37: error: 'c10d::ProcessGroupNCCL::kWatchdogThreadSleepMillis' declared 'constexpr' outside its class
Aug 21 04:05:34  constexpr int64_t ProcessGroupNCCL::kWatchdogThreadSleepMillis = 100;
Aug 21 04:05:34                                      ^
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:114:37: error: declaration of 'const int64_t c10d::ProcessGroupNCCL::kWatchdogThreadSleepMillis' outside of class is not definition [-fpermissive]
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:116:37: error: redeclaration 'c10d::ProcessGroupNCCL::kProcessGroupNCCLOpTimeoutMillis' differs in 'constexpr'
Aug 21 04:05:34  constexpr int64_t ProcessGroupNCCL::kProcessGroupNCCLOpTimeoutMillis =
Aug 21 04:05:34                                      ^
Aug 21 04:05:34 In file included from /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:1:0:
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/../c10d/ProcessGroupNCCL.hpp:197:24: error: from previous declaration 'c10d::ProcessGroupNCCL::kProcessGroupNCCLOpTimeoutMillis'
Aug 21 04:05:34    static const int64_t kProcessGroupNCCLOpTimeoutMillis;
Aug 21 04:05:34                         ^
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:116:37: error: 'c10d::ProcessGroupNCCL::kProcessGroupNCCLOpTimeoutMillis' declared 'constexpr' outside its class
Aug 21 04:05:34  constexpr int64_t ProcessGroupNCCL::kProcessGroupNCCLOpTimeoutMillis =
Aug 21 04:05:34                                      ^
Aug 21 04:05:34 /var/lib/jenkins/workspace/torch/lib/c10d/ProcessGroupNCCL.cpp:116:37: error: declaration of 'const int64_t c10d::ProcessGroupNCCL::kProcessGroupNCCLOpTimeoutMillis' outside of class is not definition [-fpermissive]
```

suo commented:
cc @zrphercule

pritamdamania87 commented:
@suo Thanks for flagging this. I'm wondering why this build wasn't run as part of the PR before I landed it? Looking at the list of tests run by CI for this PR, I see that all passed and this build wasn't included.

suo commented:
Not all variants are run on PRs. We run a subset that gives high coverage of potential breakages, but it looks like this one was unlucky.

suo commented:
Not all variants are run on PRs. We run a subset that gives high coverage of potential breakages, but it looks like this one was unlucky.

pritamdamania87 commented:
@suo I'm assuming we don't run all variants on PRs since a single PR might go through dozens of iterations causing CI to churn. Is it possible to run all variants before we finally land the PR? I'm wondering if its possible to have some variant of "pytorchbot retest this please", so that we can trigger all CI runs once before we land?

gchanan commented:
Independent of which tests run on PRs vs master, why wasn't this reverted?

suo commented:
There was no good reason not to revert. I'm reverting now
